### ISSUE TITLE
[no-floating-promises] Rule is colliding with how Reacts `useEffect` hook works

### ISSUE THREAD

**Issue by JonathanStoye** [typescript-eslint/typescript-eslint#1184]
I am aware that this is very specific, but since React as well as this pattern of React is very popular, it would be great to maybe have an option to deal with it.

**Repro**

```TS
useEffect(() => {
    const doSomethingAsync = async () => {
        try {
        await somethingAsync()
        } catch (e) {
            // ...
        }
    };
    // Here eslint complains
    doSomethingAsync();
}, []);
```

**Expected Result**
The rule is not triggered because that is the only way to use async/await inside `useEffect`
(https://reactjs.org/docs/hooks-faq.html#how-can-i-do-data-fetching-with-hooks)

**Actual Result**
The rules complains and one has to deactivate ESLint on this line.

**Additional Info**
–

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.0` |
| `@typescript-eslint/parser`        | `2.3.0` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.4.0` |
| `node`                             | `v12.12.0` |
| `npm`                              | `6.11.3` |

**Comment by bradzacher** [comment]
Just because the react docs recommend it, does not make it any less of a floating promise.

If you are okay with floating promises within your codebase (such as those within your useEffect hooks, then you should disable the rule.

**Comment by JonathanStoye** [comment]
> Just because the react docs recommend it
It is not "just" a recommendation but rather how this works.

The problem is, generally floating promises are bad but in this specific case they are not avoidable. So instead of disabling this rule and having potential floating promises all over the codebase one would rather disable this for this specific case in order to not have to disable the rule in all those places manually. Maybe with an option or something but I was hoping that since this has to be such a common problem it might be worth looking into it. But of course I might be wrong

**Comment by bradzacher** [comment]
Have a read of the docs - check out the ignoreVoid option.

It'll let you explicitly consume the promise without an eslint disable.
